File: Java/src/_0494_Target_Sum/Solution.java
package _0494_Target_Sum;

/**
 * Date: 2021/6/7 9:07
 */
public class Solution {

    int ans = 0;

    public int findTargetSumWays(int[] nums, int target) {
        dfs(nums, target, 0);
        return ans;
    }

    private void dfs(int[] nums, int target, int u) {
        if (u == nums.length) {
            if (target == 0) ans++;
            return;
        }
        dfs(nums, target - nums[u], u + 1);
        dfs(nums, target + nums[u], u + 1);
    }

    public static void main(String[] args) {

        int[] nums = {1, 1, 1, 1, 1};
        System.out.println((new Solution()).findTargetSumWays(nums, 3));  // 5
    }
}

File: Python/_0494_Target_Sum/Solution.py
# coding=utf-8
# Date: 2021/6/7 9:29
from typing import List


# 执行用时：2900 ms, 在所有 Python3 提交中击败了5.02%的用户
# 内存消耗：15.4 MB, 在所有 Python3 提交中击败了15.56%的用户
class Solution:
    def findTargetSumWays(self, a: List[int], target: int) -> int:
        if target < -1000 or target > 1000:
            return 0
        n = len(a)
        Offset = 1000
        f = [[0 for _ in range(2001)] for _ in range(n + 1)]
        f[0][Offset] = 1
        for i in range(1, n + 1):
            for j in range(-1000, 1001):
                if j - a[i - 1] >= -1000:
                    f[i][j + Offset] += f[i - 1][j - a[i - 1] + Offset]
                if j + a[i - 1] <= 1000:
                    f[i][j + Offset] += f[i - 1][j + a[i - 1] + Offset]
        return f[n][target + Offset]


if __name__ == "__main__":
    print(Solution().findTargetSumWays([1, 1, 1, 1, 1], 3))  # 5
